cupsfilters/rastertopwg.c

```diff
@@ -0,0 +1,558 @@
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+
+
+/*
+ * 'main()' - Main entry for filter.
+ */
+
+int					/* O - Exit status */
+rastertopwg(int inputfd,         /* I - File descriptor input stream */
+       int outputfd,        /* I - File descriptor output stream */
+       int inputseekable,   /* I - Is input stream seekable? (unused) */
+       filter_data_t *data, /* I - Job and printer data */
+       void *parameters)    /* I - Filter-specific parameters (unused) */
+{			/* I - Command-line arguments */
+cups_file_t	         *inputfp;		/* Print file */
+  FILE                 *outputfp;   /* Output data stream */
+  const char		*final_content_type;
+					/* FINAL_CONTENT_TYPE env var */
+  cups_raster_t		*inras,		/* Input raster stream */
+			*outras;	/* Output raster stream */
+  cups_page_header2_t	inheader,	/* Input raster page header */
+			outheader;	/* Output raster page header */
+  unsigned		y;		/* Current line */
+  unsigned char		*line;		/* Line buffer */
+  unsigned		page = 0,	/* Current page */
+			page_width,	/* Actual page width */
+			page_height,	/* Actual page height */
+			page_top,	/* Top margin */
+			page_bottom,	/* Bottom margin */
+			page_left,	/* Left margin */
+			page_right,	/* Right margin */
+			linesize,	/* Bytes per line */
+			lineoffset;	/* Offset into line */
+  int			tmp;
+  unsigned char		white;		/* White pixel */
+	/* PPD file */
+  ppd_attr_t		*back;		/* cupsBackSide attribute */
+  ppd_cache_t		*cache;		/* PPD cache */
+  pwg_size_t		*pwg_size;	/* PWG media size */
+  pwg_media_t		*pwg_media;	/* PWG media name */
+	/* Number of options */
+  cups_option_t		*options = NULL;/* Options */
+  const char		*val;		/* Option value */
+  filter_iscanceledfunc_t iscanceled = data->iscanceledfunc;
+  void                 *icd = data->iscanceleddata;
+  filter_logfunc_t     log = data->logfunc;
+  void          *ld = data->logdata;
+/*
+  * Open the input data stream specified by inputfd ...
+  */
+  
+  if ((inputfp = cupsFileOpenFd(inputfd, "r")) == NULL)
+  {
+    if (!iscanceled || !iscanceled(icd))
+    {
+      if (log) log(ld, FILTER_LOGLEVEL_ERROR,
+		   "rastertopwg: Unable to open input data stream.");
+    }
+
+    return (1);
+  }
+
+ /*
+  * Open the output data stream specified by the outputfd...
+  */
+
+  if ((outputfp = fdopen(outputfd, "w")) == NULL)
+  {
+    if (!iscanceled || !iscanceled(icd))
+    {
+      if (log) log(ld, FILTER_LOGLEVEL_ERROR,
+		   "rastertopwg: Unable to open output data stream.");
+    }
+
+    cupsFileClose(inputfp);
+
+    return (1);
+  }
+
+
+  if ((final_content_type = getenv("FINAL_CONTENT_TYPE")) == NULL)
+    final_content_type = "image/pwg-raster";
+
+  inras  = cupsRasterOpen(inputfd, CUPS_RASTER_READ);
+  outras = cupsRasterOpen(1, !strcmp(final_content_type, "image/pwg-raster") ? CUPS_RASTER_WRITE_PWG : CUPS_RASTER_WRITE_APPLE);
+
+    if (data->ppdfile == NULL && data->ppd == NULL)
+  {
+    char *p = getenv("PPD");
```

Comment:
The PPD you get via the filter data (`data->ppd`) here you also do not need to use `getenv()` and also should not use it.